### PR TITLE
Refactor Redbox, Logbox, and DevLoadingView as Sheets

### PR DESCRIPTION
**Pull request by Saadnajmi**
## Summary:

Upstream in React Native 0.74, RCTRedbox got heavily rewritten (see https://github.com/facebook/react-native/commit/e2eb26c951c3d2a0a4721428d00ffc02b501ecfa and https://github.com/facebook/react-native/commit/b311c3e0a00bf8051194bbb5072a9341cf3f6e10) . As such, we needed to update our macOS implementation. This proved to be quite difficult, as NSTableView is quite... obtuse. Many thanks to @falselobster for doing much of that part of the rewrite. I am cherry picking the rewrite commits here so we can reduce churn on the larger "0.74 merge" PR.

While we are touching this code, I realized that a natural presentation pattern for the red box would be to use a [sheet](https://developer.apple.com/design/human-interface-guidelines/sheets). In fact, that would also apply to RCTDevLoadingView (which is particularly annoying as it's presented as a fixed-in-place tiny window on your screen) and RCTLogBox. So... I went ahead and refactored all three to use sheets. The result is quite nice!

While here, I realized separately, `RCTRootView` has a `loadingView` property so you can set something like an ActivityIndicator while it loads JS. There was some modifications to get that working, So I'm also adding that in this PR. 

## Test Plan:

RCTRedBox rewritten, and presented as a sheet:

https://github.com/user-attachments/assets/f2aae788-07db-4587-9673-5084ffa057e0

RCTDevLoadingView and RCTLogBox presented as sheets:

https://github.com/user-attachments/assets/4d5e2c3d-5c1e-4b99-8a48-a0237aae030f

`RCTRootView.loadingView` set to an `NSProgressIndicator` using the code block below. Ignore the devLoadingView looking different, I was testing a different design.

```objective-c
    NSProgressIndicator *loadingView = [NSProgressIndicator new]
    [loadingView setFrame:NSMakeRect(0,0,20,20)];
    loadingView.style = NSProgressIndicatorStyleSpinning;
    [loadingView startAnimation:self];
    [((RCTRootView *)rootView) setLoadingView:loadingView];
```

https://github.com/user-attachments/assets/20c704ea-a235-4d0b-9b4a-7cf2dc02f235

### CODE REVIEW

**Review by JunielKatarn — APPROVED** [review]

File: packages/react-native/React/Base/RCTRootView.m [hunk 1]
```diff
@@ -166,20 +166,14 @@ - (void)layoutSubviews
...
-    RCTRoundPixelValue(bounds.origin.x + ((bounds.size.width - loadingViewSize.width) / 2), scale),
-    RCTRoundPixelValue(bounds.origin.y + ((bounds.size.height - loadingViewSize.height) / 2), scale)
-  );
+  CGFloat centerX = CGRectGetMidX(self.bounds);
+  CGFloat centerY = CGRectGetMidY(self.bounds);
+  NSRect newFrame = {
+    centerX - _loadingView.frame.size.width / 2.0,
+    centerY - _loadingView.frame.size.height / 2.0,
+    _loadingView.frame.size
+  };
```
Comment (amgleitman) [inline packages/react-native/React/Base/RCTRootView.m:175]: nit: Why not `NSMakeRect` to be safe?

Also, today I learned that (partial) flat initialization is a thing in C++. Who knew? 😮

File: packages/react-native/React/CoreModules/RCTDevLoadingView.mm [hunk 2]
```diff
@@ -122,8 +122,8 @@ - (void)showMessage:(NSString *)message color:(RCTUIColor *)color backgroundColo
     self->_showDate = [NSDate date];
 
     if (!self->_window && !RCTRunningInTestEnvironment()) {
+      RCTUIWindow *window = RCTKeyWindow(); // [macOS]
```
Comment (amgleitman) [inline packages/react-native/React/CoreModules/RCTDevLoadingView.mm:125]: nit: I don't see us using `window` in the macOS block. Do we need this?

File: packages/react-native/React/CoreModules/RCTDevLoadingView.mm [hunk 3]
```diff
@@ -169,7 +169,7 @@ - (void)showMessage:(NSString *)message color:(RCTUIColor *)color backgroundColo
     self->_label.textColor = color;
 
     self->_label.backgroundColor = backgroundColor;
-    [self->_window orderFront:nil];
+//    [RCTKeyWindow() beginSheet:self->_window completionHandler:nil];
```
Comment (amgleitman) [inline packages/react-native/React/CoreModules/RCTDevLoadingView.mm:172]: nit: Remove?

**Review by amgleitman — APPROVED** [review]

File: packages/react-native/React/CoreModules/RCTDevLoadingView.mm (hunk 2 above)
Comment (Saadnajmi) [reply packages/react-native/React/CoreModules/RCTDevLoadingView.mm:125]: Ah yeah, latest iteration doesn't. Removing.

**Pull request merged**